homeassistant/components/aquacell/sensor.py

```diff
@@ -0,0 +1,122 @@
+from dataclasses import dataclass
+
+from aioaquacell import Softener
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.const import PERCENTAGE, UnitOfTime
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from . import AquacellConfigEntry
+from .coordinator import AquacellCoordinator
+from .entity import AquacellEntity
+
+PARALLEL_UPDATES = 1
+
+
+@dataclass(frozen=True, kw_only=True)
+class SoftenerSensorEntityDescription(SensorEntityDescription):
+    """Describes Softener sensor entity."""
+
+    value_fn: Callable[[Softener], StateType]
+
+
+SENSORS: tuple[SoftenerSensorEntityDescription, ...] = (
+    SoftenerSensorEntityDescription(
+        key="salt_left_side_percentage",
+        translation_key="salt_left_side_percentage",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: softener.salt.leftPercent,
+    ),
+    SoftenerSensorEntityDescription(
+        key="salt_right_side_percentage",
+        translation_key="salt_right_side_percentage",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: softener.salt.rightPercent,
+    ),
+    SoftenerSensorEntityDescription(
+        key="salt_left_side_remaining_days",
+        translation_key="salt_left_side_remaining_days",
+        device_class=SensorDeviceClass.DURATION,
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.leftDays,
+    ),
+    SoftenerSensorEntityDescription(
+        key="salt_right_side_remaining_days",
+        translation_key="salt_right_side_remaining_days",
+        device_class=SensorDeviceClass.DURATION,
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.rightDays,
+    ),
+    SoftenerSensorEntityDescription(
+        key="battery",
+        device_class=SensorDeviceClass.BATTERY,
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: softener.battery,
+    ),
+    SoftenerSensorEntityDescription(
+        key="wifi_level",
+        translation_key="wifi_level",
+        value_fn=lambda softener: softener.wifiLevel,
+        device_class=SensorDeviceClass.ENUM,
+        options=[
+            "high",
+            "medium",
+            "low",
+        ],
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: AquacellConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the sensors."""
+    for idx, _ in enumerate(config_entry.runtime_data.data):
+        async_add_entities(
+            SoftenerSensor(config_entry.runtime_data, sensor, idx) for sensor in SENSORS
+        )
```

Comment:
Please collect the entities in a list before passing them to the async_add_entities. Or use a generator expression